ratel/src/ast/function.rs

```diff
@@ -69,6 +69,13 @@ pub struct Function<'ast, N: Name<'ast>> {
     pub body: BlockNode<'ast, Statement<'ast>>,
 }
 
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub struct GeneratorFunction<'ast, N: Name<'ast>> {
```

Comment:
To keep things dry, but not having to waste performance on boolean checks, this can be done by introducing a static flag by generics on the main `Function` struct. That is a bit of more advanced rust, but it would be something like this:

```rust
trait Flag {}

struct True;
struct False;

impl Flag for True {}
impl Flag for False {}

#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Function<'ast, N: Name<'ast>, G: Flag> {
    pub name: N,
    pub generator: G,
    pub params: PatternList<'ast>,
    pub body: BlockNode<'ast, Statement<'ast>>,
}
```

This is equivalent in performance to having a separate struct definition for `GeneratorFunction`, but more DRY (the compiler will unwind it to 3 (for each N) * 2 (for each G) = 6 variants on it's own). We could use the same scheme for `async` functions (making it 12 static variants no extra cost on serializing). 

All that said, all of this might be over-engineering and I think just adding a `bool` here instead is totally fine :).

Reply:
:arrow_up: the more I think about it, the more I think just using a `bool` is the best, especially since we also want to allow `async` :).

Reply:
I agree, a signature like the following makes more sense and follows the DRY principle.

````rust
#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Function<'ast, N: Name<'ast>> {
    pub name: N,
    pub params: PatternList<'ast>,
    pub body: BlockNode<'ast, Statement<'ast>>,
    pub generator: bool,
    pub async: bool,
}
````